### PR TITLE
CFE-2240: Added policy function `type()`

### PR DESCRIPTION
**Pull request by larsewi**
Ticket: 2240
Changelog: Title
Signed-off-by: larsewi <lars.erik.wik@northern.tech>

### CODE REVIEW

**Comment by larsewi** [comment]
I think the name `type` makes more sense than `datatype`. Please let me know if you disagree.

File: libpromises/evalfunction.c [hunk 1]
```diff
@@ -5119,69 +5119,85 @@ static FnCallResult FnCallFold(EvalContext *ctx,
...
+    }
 
-        const JsonElement* const jelement = value;
+    if (type == CF_DATA_TYPE_CONTAINER)
+    {
+        const char *subtype_str;
+        const JsonElement *const element = value;
 
-        if (JsonGetElementType(jelement) == JSON_ELEMENT_TYPE_CONTAINER)
+        if (JsonGetElementType(element) == JSON_ELEMENT_TYPE_CONTAINER)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5161]: @larsewi Just use [`JsonGetType()`](https://github.com/cfengine/libntech/blob/8a22d70481deb4f93e41a1b58c416e0236b6a275/libutils/json.h#L184) instead, then you can make one big switch for [all the types](https://github.com/cfengine/libntech/blob/8a22d70481deb4f93e41a1b58c416e0236b6a275/libutils/json.h#L71-L80).

**Review by olehermanse — CHANGES_REQUESTED** [review]

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -5119,69 +5119,85 @@ static FnCallResult FnCallFold(EvalContext *ctx,
 }
 
 /*********************************************************************/
-/* This function has been removed from the function list for now     */
-/*********************************************************************/
-#ifdef SUPPORT_FNCALL_DATATYPE
+
 static FnCallResult FnCallDatatype(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5123]:
```suggestion
static FnCallResult FnCallType(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
```

Comment (larsewi) [reply libpromises/evalfunction.c:5123]: @olehermanse the name `FnCallType` is already used. Please see FnCall.h:78.

File: tests/acceptance/01_vars/02_functions/type.cf [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      "test_29" string => type("init._rlist", "on");
+
+      "test_30" string => type("undefined");
+      "test_31" string => type("undefined", "true");
+
+      "test_32" string => type(or("any", "!any"));
+      "test_33" string => type(or("any", "!any"), "true");
+
+      "test_34" string => type(and("any", "!any"));
+      "test_35" string => type(and("any", "!any"), "true");
```
Comment (olehermanse) [inline tests/acceptance/01_vars/02_functions/type.cf:89]: @larsewi Sorry, I keep forgetting that this function should take a variable name. Then this doesn't make much sense. Please remove it again. 😭 

```suggestion
```

**Review by olehermanse — CHANGES_REQUESTED** [review]
Please remove the `and()` `or()` as suggested above.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6554/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6554/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6554/

**Pull request merged**